src/mpid/ch4/generic/am/mpidig_am_msg.h

```diff
@@ -199,15 +200,17 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
                             MPIDIG_REQUEST(rreq, buffer),
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), p->offset, &actual_unpack_bytes);
-        p->in_data_sz -= actual_unpack_bytes;
-        p->offset += actual_unpack_bytes;
         if (!rreq->status.MPI_ERROR && payload_sz > actual_unpack_bytes) {
             /* basic element size mismatch */
             rreq->status.MPI_ERROR =
                 MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
                                      MPI_ERR_TYPE, "**dtypemismatch", 0);
-            return 1;
-        } else if (p->in_data_sz == 0) {
+        }
+
+        p->in_data_sz -= payload_sz;
+        p->offset += payload_sz;
+
+        if (p->in_data_sz == 0) {
```

Comment:
This part doesn't belong to this patch.

Reply:
Wait, this is the only part in this commit. Earlier we assume `actual_unpack_bytes` is always the same as `payload_sz` unless it is "truncation error", which we abort. This patch fixes it due to the understanding that `actual_unpack_bytes` may not correlate to actual "remaining in_data_sz" or `offset`, so we should directly adjust it by `payload_sz` and check returning status by whether "remaining in_data_sz" is 0.

Reply:
Using `payload_sz` is incorrect here.  `actual_unpack_bytes` was the correct amount of data moved to the user buffer.

Reply:
> Using `payload_sz` is incorrect here. `actual_unpack_bytes` was the correct amount of data moved to the user buffer.

That's true, but `p->in_data_sz` is tracking bytes coming from the sender, not necessarily what is being placed into the recv buffer. Perhaps we should be careful about setting the offset value such that it doesn't cause an out-of-bounds access, though.